prov/efa/src/rxr/rxr_pkt_cmd.c

```diff
@@ -1055,6 +1055,116 @@ void rxr_pkt_proc_received(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_entry)
 	}
 }
 
+/**
+ * @brief Read peer raw address from packet header and insert the peer in AV.
+ * The caller of this function must guarantee that the packet header contain the raw address.
+ * @param ep Pointer to RDM endpoint
+ * @param pkt_entry Pointer to packet entry
+ * @returns Peer address
+ */
+static inline fi_addr_t rxr_pkt_determine_addr_from_hdr(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_entry)
+{
+	void *raw_addr;
+
+	raw_addr = rxr_pkt_req_raw_addr(pkt_entry);
+	assert(raw_addr);
+	return rxr_pkt_insert_addr(ep, pkt_entry, raw_addr);
+};
+
+
+#if HAVE_EFADV_CQ_EX
+/**
+ * @brief Read peer raw address from EFA device and look up the peer address in AV.
+ * The caller must guarantee that the peer AH is unknown.
+ * @return Peer address, or FI_ADDR_NOTAVAIL if unavailable.
+ */
+static inline fi_addr_t rxr_pkt_determine_address_from_efadv(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_entry)
+{
+	struct efa_cq *efa_cq;
+	struct efa_ep *efa_ep;
+	struct efa_ep_addr efa_ep_addr = {0};
+	fi_addr_t addr;
+	union ibv_gid gid = {0};
+	uint32_t *connid = NULL;
+
+	efa_cq = container_of(ep->rdm_cq, struct efa_cq, util_cq.cq_fid);
+
+	/* Attempt to read AH from EFA firmware */
+	if (efadv_wc_read_ah(efadv_cq_from_ibv_cq_ex(efa_cq->ibv_cq_ex), &gid) >= 0) {
```

Comment:
> but there is no guarantee of it.

Could you explain with an example? Currently this code path is inline with `rdm_ep_poll_ibv_cq` and we make sure `efadv_wc_read_ah` is called before `ibv_next_poll`.

As we discussed earlier, moving this call higher up the call stack could introduce overhead, since we will essentially 1) parse packet header and 2) query efadv before looking up reverse AV.

Reply:
Yes. Today it is correct. but if someone change the call sequence in `rdm_ep_poll_ibv_cq` in the future. it became incorrect, and will the bug will be hard to debug, because `efadv_wc_read_ah` can still return correct value.

> As we discussed earlier, moving this call higher up the call stack could introduce overhead, since we will essentially 1) parse packet header and 2) query efadv before looking up reverse AV.

I think we can call this after `efa_av_lookup_address_rdm` return FI_ADDR_NOTAVAIL. Then we maintain the same logic.

Reply:
> I suggest testing the function call ordering and make sure efadv_wc_read_ah is always called before ibv_next_poll.

This will not be true if we ever pull 2 CQ entries from ibv_cq.

> but I think the current code structure is cleaner.

The change to `rdm_ep_poll_ibv_cq` should be small too:

After 
```
    pkt_entry->addr = efa_av_lookup_addr_rdm(...)
```

add two lines:
```
if (pkt_entry->addr == FI_ADDR_NOTAVAIL)
    pkt_entry->addr = rxr_ep_determine_addr_from_ibv_cq_ex(rxr_ep, ibv_cq_ex);
```
> Rather than changing the critical path

Not the sure I understand, current change is also in critical path.

Reply:
Addressed in new revision. Moved things around `rxr_ep.c` and `rxr_pkt_cmd.c`, such that packet address is determined before recv completion call.